### ISSUE TITLE
Cmdlets over cron and task scheduler

### ISSUE THREAD

**Issue by SteveL-MSFT** [PowerShell/PowerShell#7541]
This came out of today's Community Call.  Summary of the discussion is that:

  - cron and task scheduler are fundamentally different in syntax and semantics that an abstraction would be highly complex and probably still incomplete
  - makes more sense to focus on the top scenarios and build a new set of cmdlets to enable scheduling cross-platform

So this issue is to brainstorm on requirements and scenarios before we move towards authoring a RFC.

**Comment by DarwinJS** [comment]
One possible guiding principle is that since cron is likely a subset of schedule tasks, that we start with full cron - but remove any agreed esoterics.

**Comment by SteveL-MSFT** [comment]
@DarwinJS it would be great if we had some specific examples of some of the capabilities required from existing use of cron and existing use of task scheduler

**Comment by thezim** [comment]
On macOS will use `launchd` or `cron`. The later has been deprecated since 10.11 if recall. It is still there but in a disabled state and no files exist for it by default.

**Comment by DarwinJS** [comment]
@SteveL-MSFT - here are some examples of how I do use scheduling.  I can't say that they are the uber example - but they are ones that would benefit from CMDLet parity.

Also the main focus in the below is a) server b) automation - as such it exhibits using the scheduler for things like "getting into SYSTEM account context (windows)" and "scheduling something for shutdown or startup" and "scheduling something to run only once"

This article directly talks about the overcomplexity of the current CMDLets and why schtasks.exe is better.  It shows the comparison of the code complexity:
https://cloudywindows.io/post/continue-your-automation-to-run-once-after-restarting-a-headless-windows-system

Notice it takes advantage of a super easy way to get admin without a password (assuming you run the schedule command with admin) = "/ru SYSTEM" and the ability to run on startup "/sc ONSTART"

This code does "self-unscheduling".

This brings up a positive distinction about Windows well developed scheduler over linux - it has these sort of meta-functions that a) aren't strictly the typical domain of a pure scheduler, b) that are some of it's most useful shortcuts.  Stuff like ONSTART and SYSTEM account without admin.  If this concept could be extended to linux - it would make the PowerShell version of scheduling legions better than the native.  I know it is a big ask - but I've always loved the way PowerShell surfaces super-needed, super-simple high level functionality like this.

These two scripts are meant to have identical functionality on Windows and Linux.  They use a repeating task to: *) run in parallel to remaining automation, *) persist across reboots that are so common during initial automation:
In Windows I used schtasks.exe to avoid the complexity and limited availability of scheduling cmdlets:
https://github.com/DarwinJS/DevOpsAutomationCode/blob/master/InitializeDisksWithFIO.ps1#L274
https://github.com/DarwinJS/DevOpsAutomationCode/blob/master/InitializeDisksWithFIO.sh#L343
This code does "self-unscheduling".

Here's an example of running a cleanup.  The CMDLets were used because I only needed one.  Surprising running elevated didn't take more cmdlets.

**Running with root / system w/ no password would be a feature to be on a simple cmdlet because it is used so frequently in server automation**

Here is a use case of using scheduler to do something on the next start (rather than fussing with each systems particular "startup automation" techniques:
https://github.com/DarwinJS/ChocoPackages/blob/master/PowerShell/v5.1/tools/ChocolateyInstall.ps1#L168

**Running only once and running on events like startup or shutdown would be very handy as those are frequent too** - however I understand they aren't implicitly a feature in both schedulers.

**Self-unscheduling upon completion** - is another thing I use time and again for system automation.

I have resorted to the scheduler on Windows for so many more scenarios that simply "setup a repeating scheduled job" - that I guess it would seem a crime to roll back new functionality to only be that.  I think that newly presented functionality might want to acknowledge that on a bare bones Windows or Linux system, the scheduler has other identifable defacto "meta uses" patterns.

Multi-platform functionality that would be useful to my work (ignoring how "hard" they might be):
- way to gain SYSTEM / root account context (without knowing implementation details)
- easy way to setup code to run on system wide events like startup, shutdown (without knowing implementation details)
- way to self-remove a task upon successful completion (without knowing implementation details)
- simple way to re-run, re-run, re-run until the job is done.  Maybe scheduler function takes a flag file name to know when it is done.  Almost always combined with self-remove.

I am already using PowerShell jobs on both platforms to run in parallel - but jobs are not reboot nor session resilient.

Also - sometimes implementations like the above (scheduler "meta-functions") get labeled esoteric.  My counter to that is "If we can guess that these things would receive wide adoption upon recognition by potential users - then they are only esoteric in a short term sense - the sense also described by 'innovating something that customers can't articulate before they see it - but wholehearted recognize when they do'"   In other words "we've never heard anyone do that or ask for that before" is not an exclusive vote for "esoteric and unneeded" it could just as easily be "unarticulated and unrecognized, yet very needed"  

To me, the entire story of PowerShell is actually one that carries the same sense - I'm interested in the continuance of that spirit rather than devolving to "minimum common functionality" just because doing multiplatform is hard.

**Comment by jdhitsolutions** [comment]
Added to the Cmdlets Working Group for review.

**Comment by jdhitsolutions** [comment]
The Cmdlet Working Group has discussed this issue. The PSScheduledJob module is not supported in PowerShell 7 because .NET Core no longer supports `System.Runtime.Serialization.NetDataContractSerializer`. This is the primary reason this is marked as `won't fix`. The group also believes that while the Windows task scheduler and Unix cron,  do the same thing conceptually are architecturally different enough that building a cmdlet set to abstract the process over both is not practical. It would be better to have different sets of cmdlets. 

In Windows, PowerShell 7 supports the ScheduledTasks module. Users could create PowerShell scheduled tasks with these cmdlets today. The Working Group is recommending that any solution to creating PowerShell scheduled jobs be left to the community and published to the PowerShell Gallery.
